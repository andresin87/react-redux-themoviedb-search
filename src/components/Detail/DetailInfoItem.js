import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { List } from 'antd';

class DetailInfoItem extends PureComponent {
  getNamesOfArray = array => {
    return array
      .reduce((namesOfArray, item) => {
        return item.name ? `${namesOfArray} ${item.name},` : '';
      }, '')
      .slice(0, -1);
  };

  getDateWithDDMMYYYFormat = date => {
    var pattern = /(\d{4})-(\d{2})-(\d{2})/;
    if (!date || !date.match(pattern)) {
      return null;
    }
    return date.replace(pattern, '$3/$2/$1');
  };

  getItemContent = item => {
    const { type, value } = item;

    switch (type) {
      case 'array':
        return `${value && value.length ? this.getNamesOfArray(value) : ''}`;
      case 'currency':
        return `${value ? value.toLocaleString() : ''}$`;
      case 'date':
        return `${value ? this.getDateWithDDMMYYYFormat(value) : ''}`;
      case 'number':
        return `${value ? value.toLocaleString() : ''}`;
      case 'time':
        return `${value ? value + ' min.' : ''}`;
      case 'text':
      default:
        return `${value ? value : ''}`;
    }
  };

  render() {
    const { children } = this.props;
    const { label, value } = children;
    if (!value || value === '') return null;
    return (
      <List.Item>
        <strong style={{ whiteSpace: 'nowrap', minWidth: '140px' }}>
          {label}
          :&nbsp;
        </strong>
        {this.getItemContent(children)}
      </List.Item>
    );
  }
}

DetailInfoItem.propTypes = {
  key: PropTypes.string,
  label: PropTypes.string,
  type: PropTypes.oneOf([
    'currency',
    'array',
    'text',
    'number',
    'date',
    'time',
  ]),
  value: PropTypes.string,
};

export default DetailInfoItem;
