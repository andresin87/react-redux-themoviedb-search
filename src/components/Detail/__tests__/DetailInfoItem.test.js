import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import DetailInfoItem from '../DetailInfoItem';

const dataSource = [
  {
    key: 'budget',
    label: 'Budget',
    type: 'currency',
    value: 63000000,
  },
  {
    key: 'genres',
    label: 'Genres',
    type: 'array',
    value: [
      {
        id: 18,
        name: 'Drama',
      },
    ],
  },
  {
    key: 'overview',
    label: 'Overview',
    type: 'text',
    value:
      'A ticking-time-bomb insomniac and a slippery soap salesman channel primal male aggression into a shocking new form of therapy. Their concept catches on, with underground "fight clubs" forming in every town, until an eccentric gets in the way and ignites an out-of-control spiral toward oblivion.',
  },
  {
    key: 'popularity',
    label: 'Popularity',
    type: 'number',
    value: 38.513,
  },
  {
    key: 'production_companies',
    label: 'Prod. companies',
    type: 'array',
    value: [
      {
        id: 508,
        logo_path: '/7PzJdsLGlR7oW4J0J5Xcd0pHGRg.png',
        name: 'Regency Enterprises',
        origin_country: 'US',
      },
      {
        id: 711,
        logo_path: '/tEiIH5QesdheJmDAqQwvtN60727.png',
        name: 'Fox 2000 Pictures',
        origin_country: 'US',
      },
      {
        id: 20555,
        logo_path: null,
        name: 'Taurus Film',
        origin_country: '',
      },
      {
        id: 54051,
        logo_path: null,
        name: 'Atman Entertainment',
        origin_country: '',
      },
      {
        id: 54052,
        logo_path: null,
        name: 'Knickerbocker Films',
        origin_country: '',
      },
      {
        id: 25,
        logo_path: '/qZCc1lty5FzX30aOCVRBLzaVmcp.png',
        name: '20th Century Fox',
        origin_country: 'US',
      },
      {
        id: 4700,
        logo_path: '/A32wmjrs9Psf4zw0uaixF0GXfxq.png',
        name: 'The Linson Company',
        origin_country: '',
      },
    ],
  },
  {
    key: 'production_countries',
    label: 'Prod. countries',
    type: 'array',
    value: [
      {
        iso_3166_1: 'DE',
        name: 'Germany',
      },
      {
        iso_3166_1: 'US',
        name: 'United States of America',
      },
    ],
  },
  {
    key: 'release_date',
    label: 'Release date',
    type: 'date',
    value: '1999-10-15',
  },
  {
    key: 'runtime',
    label: 'Runtime',
    type: 'time',
    value: 139,
  },
  {
    key: 'spoken_languages',
    label: 'Spoken languages',
    type: 'array',
    value: [
      {
        iso_639_1: 'en',
        name: 'English',
      },
    ],
  },
  {
    key: 'status',
    label: 'Status',
    type: 'text',
    value: 'Released',
  },
  {
    key: 'vote_average',
    label: 'Vote average',
    type: 'number',
    value: 8.4,
  },
  {
    key: 'vote_count',
    label: 'Vote count',
    type: 'number',
    value: 13653,
  },
];

describe('DetailInfoItem component', () => {
  dataSource.forEach(item => {
    describe(`Processing ${item.key}`, () => {
      const component = <DetailInfoItem>{item}</DetailInfoItem>;

      it(`Renders ${item.key}`, () => {
        const div = document.createElement('div');
        ReactDOM.render(component, div);
        ReactDOM.unmountComponentAtNode(div);
      });

      it(`Snapshot ${item.key}`, () => {
        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot(`Item ${item.key}`);
      });
    });
  });
});
