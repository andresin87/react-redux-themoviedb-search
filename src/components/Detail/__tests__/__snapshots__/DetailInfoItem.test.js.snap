// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item budget 1`] = `
<mocked
  name="ListItem"
  props={Object {}}
>
  <strong
    style={
      Object {
        "minWidth": "140px",
        "whiteSpace": "nowrap",
      }
    }
  >
    Budget
    : 
  </strong>
  63,000,000$
</mocked>
`;

exports[`Item genres 1`] = `
<mocked
  name="ListItem"
  props={Object {}}
>
  <strong
    style={
      Object {
        "minWidth": "140px",
        "whiteSpace": "nowrap",
      }
    }
  >
    Genres
    : 
  </strong>
   Drama
</mocked>
`;

exports[`Item overview 1`] = `
<mocked
  name="ListItem"
  props={Object {}}
>
  <strong
    style={
      Object {
        "minWidth": "140px",
        "whiteSpace": "nowrap",
      }
    }
  >
    Overview
    : 
  </strong>
  A ticking-time-bomb insomniac and a slippery soap salesman channel primal male aggression into a shocking new form of therapy. Their concept catches on, with underground "fight clubs" forming in every town, until an eccentric gets in the way and ignites an out-of-control spiral toward oblivion.
</mocked>
`;

exports[`Item popularity 1`] = `
<mocked
  name="ListItem"
  props={Object {}}
>
  <strong
    style={
      Object {
        "minWidth": "140px",
        "whiteSpace": "nowrap",
      }
    }
  >
    Popularity
    : 
  </strong>
  38.513
</mocked>
`;

exports[`Item production_companies 1`] = `
<mocked
  name="ListItem"
  props={Object {}}
>
  <strong
    style={
      Object {
        "minWidth": "140px",
        "whiteSpace": "nowrap",
      }
    }
  >
    Prod. companies
    : 
  </strong>
   Regency Enterprises, Fox 2000 Pictures, Taurus Film, Atman Entertainment, Knickerbocker Films, 20th Century Fox, The Linson Company
</mocked>
`;

exports[`Item production_countries 1`] = `
<mocked
  name="ListItem"
  props={Object {}}
>
  <strong
    style={
      Object {
        "minWidth": "140px",
        "whiteSpace": "nowrap",
      }
    }
  >
    Prod. countries
    : 
  </strong>
   Germany, United States of America
</mocked>
`;

exports[`Item release_date 1`] = `
<mocked
  name="ListItem"
  props={Object {}}
>
  <strong
    style={
      Object {
        "minWidth": "140px",
        "whiteSpace": "nowrap",
      }
    }
  >
    Release date
    : 
  </strong>
  15/10/1999
</mocked>
`;

exports[`Item runtime 1`] = `
<mocked
  name="ListItem"
  props={Object {}}
>
  <strong
    style={
      Object {
        "minWidth": "140px",
        "whiteSpace": "nowrap",
      }
    }
  >
    Runtime
    : 
  </strong>
  139 min.
</mocked>
`;

exports[`Item spoken_languages 1`] = `
<mocked
  name="ListItem"
  props={Object {}}
>
  <strong
    style={
      Object {
        "minWidth": "140px",
        "whiteSpace": "nowrap",
      }
    }
  >
    Spoken languages
    : 
  </strong>
   English
</mocked>
`;

exports[`Item status 1`] = `
<mocked
  name="ListItem"
  props={Object {}}
>
  <strong
    style={
      Object {
        "minWidth": "140px",
        "whiteSpace": "nowrap",
      }
    }
  >
    Status
    : 
  </strong>
  Released
</mocked>
`;

exports[`Item vote_average 1`] = `
<mocked
  name="ListItem"
  props={Object {}}
>
  <strong
    style={
      Object {
        "minWidth": "140px",
        "whiteSpace": "nowrap",
      }
    }
  >
    Vote average
    : 
  </strong>
  8.4
</mocked>
`;

exports[`Item vote_count 1`] = `
<mocked
  name="ListItem"
  props={Object {}}
>
  <strong
    style={
      Object {
        "minWidth": "140px",
        "whiteSpace": "nowrap",
      }
    }
  >
    Vote count
    : 
  </strong>
  13,653
</mocked>
`;
